### ISSUE TITLE
[CT-3546] Remove restrictive `pathspec` version depencency (upper bound)

### ISSUE THREAD

**Issue by rzjfr** [dbt-labs/dbt-core#9372]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

Currently [pathspec>=0.9,<0.12](https://github.com/dbt-labs/dbt-core/blob/main/core/setup.py#L68) which conflicts with latest version of pathspec (0.12.1). I can't find any reason specifically for <0.12. It seems to be consequence of dependebot upgrading more restrictive pathspec~=0.9 [here](https://github.com/dbt-labs/dbt-core/pull/5917) and then later [here](https://github.com/dbt-labs/dbt-core/pull/6737)

### Expected Behavior

I can't find any reason to restrict pathspec <0.12. I couldn't find any major change that can break anything related to this package in [pathspec change log](https://github.com/cpburnz/python-pathspec/blob/v0.12.1/CHANGES.rst).

### Steps To Reproduce

1. `pip install dbt-core==1.7.4 -c <(echo "pathspec==0.12.1")`

### Relevant log output

```shell
The conflict is caused by:
    dbt-core 1.7.4 depends on pathspec<0.12 and >=0.9
    The user requested (constraint) pathspec==0.12.1
```

### Environment

```markdown
- OS: NA
- Python: >3.8
- dbt: >1.2.6
```

**Comment by WillAyd** [comment]
Looking forward to the fix. Airflow 2.8 pins pathspec to 0.12.1 so it is impossible to use the current 1.7.x series there without violating constraints

https://raw.githubusercontent.com/apache/airflow/constraints-2.8.1/constraints-3.11.txt

**Comment by safwank** [comment]
Any chance this will be merged soon? We recently upgraded Airflow to 2.8 and ran into the same issue. Let me know if there's anything I can do to make it happen.
